osu.Server.Queues.ScoreStatisticsProcessor/Commands/Queue/BatchInserter.cs

```diff
@@ -379,6 +373,36 @@ private static LegacyBeatmapConversionDifficultyInfo getDificultyInfo(int beatma
             }
         }
 
+        private static readonly ConcurrentDictionary<int, RulesetCache> ruleset_cache = new ConcurrentDictionary<int, RulesetCache>();
+
+        private static RulesetCache getRulesetCache(int rulesetId)
+        {
+            if (ruleset_cache.TryGetValue(rulesetId, out var cache))
+                return cache;
+
+            var ruleset = LegacyRulesetHelper.GetRulesetFromLegacyId(rulesetId);
+
+            HitResult maxBasicResult;
+            using (var scoreProcessor = ruleset.CreateScoreProcessor())
+                maxBasicResult = ruleset.GetHitResults().Where(h => h.result.IsBasic()).Select(h => h.result).MaxBy(scoreProcessor.GetBaseScoreForResult);
+
+            cache = new RulesetCache(ruleset, maxBasicResult, ruleset.CreateMod<ModClassic>()!);
+
+            ruleset_cache[rulesetId] = cache;
+
+            return cache;
+        }
+
+        private record RulesetCache(Ruleset Ruleset, HitResult MaxBasicResult, Mod ClassicMod);
```

Comment:
`record`? in _your_ code? well this is an unexpected development :smile:

Reply:
I was just matching suit below. 🤷